### PR TITLE
Exp docs 0.2.0 preview

### PR DESCRIPTION
**Pull request by laserkelvin**
# ALCHEMI Toolkit Pull Request

## Description

This PR revamps the documentation for the 0.2.0 release, including user-facing and API documentation across the board

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

-
-
-

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [ ] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [ ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ ] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/training/hooks/finetune.py [hunk 1]
```diff
@@ -216,32 +207,6 @@ class TrainableParameterHook(BaseModel):
     ``freeze_mode="optimizer_only"`` to preserve gradients for excluded
     parameters while keeping them out of optimizer parameter groups.
```
Comment (ys-teh) [inline nvalchemi/training/hooks/finetune.py:217]: It'd be useful to add the following for clarity:

"""The hook computes a fully-qualified parameter allow-list when registered
    on a :class:`~nvalchemi.training.strategy.TrainingStrategy`. Explicit
    trainable selections without ``freeze_patterns`` form an allow-list. When
    ``freeze_patterns`` are also supplied, matching parameters are excluded
    first, then explicit trainable selections are re-included. Parameters
    outside ..."""

Comment (laserkelvin) [reply nvalchemi/training/hooks/finetune.py:217]: I added a variation of that with the intention of making it more easily scannable. Let me know if you'd prefer your text

File: docs/userguide/training.md [hunk 2]
```diff
@@ -74,79 +80,75 @@ single diagram below, which is useful as a reference when you are trying
...
+     after_loss -> before_backward -> do_backward -> after_backward;
+     after_backward -> before_step -> do_step -> after_step -> after_batch;
+     after_batch -> before_batch [label="next batch", style=dashed];
+     after_batch -> after_epoch [label="epoch exhausted"];
+     after_epoch -> before_epoch [label="next epoch", style=dashed];
+     after_step -> after_validation [label="every_n_steps", style=dotted];
+     after_epoch -> after_validation [label="every_n_epochs", style=dotted];
+     after_epoch -> after_training [label="target reached"];
+     after_training -> final_validation -> after_validation;
+   }
```
Comment (ys-teh) [inline docs/userguide/training.md]: This block doesn't render well on my end. Check it.

Image

File: docs/userguide/reporting.md [hunk 3]
```diff
@@ -63,31 +63,27 @@ is to act with the requested cadence on each provided reporter:
...
+       state        [label="ReportingState\n(event metadata + recent messages)"]
+       reporter     [label="Reporter\n(TensorBoard / Rich / custom)"]
+       output       [label="Output\n(file / run log / dashboard)"]
+
+       workflow -> context     [label="engine hook call"]
+       context -> orchestrator [label="stage match"]
+       orchestrator -> state   [label="mark event"]
+       state -> reporter       [label="report"]
+       reporter -> output      [label="write"]
+   }
```
Comment (ys-teh) [inline docs/userguide/reporting.md]: This block doesn't render well for me as well. Maybe it's due to my rendering settings?

File: docs/userguide/reporting.md
Comment (ys-teh) [inline docs/userguide/reporting.md:193]: If it is not too much work, it would be nice to provide a couple of screenshots to show how the reporter looks like, for example, one using training layout and the dynamics layout. This would help to attract users to invest their time in building one.

File: docs/userguide/distributed.md
Comment (ys-teh) [inline docs/userguide/distributed.md]: minimal end-to-end usage example?

File: docs/userguide/distributed_design.md
Comment (ys-teh) [inline docs/userguide/distributed_design.md]: This looks the same as the above block code?

Comment (ys-teh) [inline docs/userguide/distributed_design.md]: Fix the numbering here and the subsequent ones.

**Review by ys-teh — COMMENTED** [review]
Skimmed through a few pages, and everything looks fine to me so far. Additional visual would be helpful in general, though some of the graphs and flowcharts may just not have rendered correctly on my end. I also left some minor comments around.

File: nvalchemi/training/strategy.py [hunk 4]
```diff
@@ -277,68 +277,113 @@ def default_training_fn(model: BaseModelMixin, batch: Batch) -> dict[str, torch.
...
+    ...     models=model,
+    ...     optimizer_configs=OptimizerConfig(optimizer_cls=torch.optim.AdamW),
+    ...     num_steps=50_000,
+    ...     training_fn=default_training_fn,
+    ...     loss_fn=EnergyMSELoss(),
+    ...     validation_config=ValidationConfig(
+    ...         validation_data=val_batches,
+    ...         every_n_steps=1_000,
+    ...     ),
+    ...     hooks=[CheckpointHook(root_folder="runs/exp")],
```
Comment (ys-teh) [inline nvalchemi/training/strategy.py]: CheckpointHook takes checkpoint_dir, not root_folder.

File: docs/userguide/agent_skills.md [hunk 5]
```diff
@@ -28,6 +28,9 @@ your project's skills directory, or your user-level one (e.g.
 | `nvalchemi-data-storage` | How to write, read, compose, and load atomic data using the composable Zarr-backed storage pipeline (Writer, Reader, Dataset, MultiDataset, DataLoader). | {ref}`datapipes_guide` |
 | `nvalchemi-zarr-perf` | How to tune Zarr-backed Reader, Dataset, MultiDataset, and DataLoader throughput with fused reads, validation skipping, pinned memory, and benchmark sweeps. | {ref}`read_performance_tuning` |
 | `nvalchemi-model-wrapping` | How to wrap an arbitrary MLIP using the {py:class}`~nvalchemi.models.base.BaseModelMixin` interface to standardize inputs, outputs, and embeddings. | {ref}`models_guide` |
+| `nvalchemi-training-api` | How to configure, run, extend, and debug {py:class}`~nvalchemi.training.TrainingStrategy` workflows. | {ref}`training_guide` |
```
Comment (ys-teh) [inline docs/userguide/agent_skills.md:31]: Duplicated rows. Please delete lines 37 to 39.

**Review by ys-teh — CHANGES_REQUESTED** [review]

File: nvalchemi/training/strategy.py [hunk 6]
```diff
@@ -1366,6 +1543,21 @@ def save_checkpoint(
...
         """Save this strategy as a restartable checkpoint.
 
+        Rather than pickling the strategy, this writes a spec-based checkpoint:
+        model weights and architecture, optimizer and scheduler state, the
+        training counters (``step_count``, ``epoch_count``, ``batch_count``,
+        ``global_step_count``), and the state of any
+        :class:`~nvalchemi.hooks.CheckpointableHook` are each serialized through
+        their Pydantic specs, so a restart reconstructs the objects without
+        executing arbitrary pickled code. The non-serializable pieces --
+        ``training_fn`` and ``loss_target_assembler`` -- are intentionally
```
Comment (ys-teh) [inline nvalchemi/training/strategy.py:1553]: Future consideration, not necessarily a blocker for this PR: it may be useful to document the general rule for what can be serialized vs what must be supplied again at load time. That would likely make the behavior more intuitive for users and reduce trial-and-error.

File: docs/userguide/training.md (hunk 2 above)
Comment (laserkelvin) [reply docs/userguide/training.md]: You would need `graphviz` to render it

File: docs/userguide/reporting.md (hunk 3 above)
Comment (laserkelvin) [reply docs/userguide/reporting.md]: You need `graphviz`, which is unfortunately not something you can `pip` install

File: nvalchemi/training/strategy.py (hunk 6 above)
Comment (laserkelvin) [reply nvalchemi/training/strategy.py:1553]: Added in 58db5cb and [a8bec0d](https://github.com/NVIDIA/nvalchemi-toolkit/pull/146/commits/a8bec0d3ae65074086340512f932586a883f91da)

File: docs/userguide/reporting.md
Comment (laserkelvin) [reply docs/userguide/reporting.md:193]: Added in 9c901db

**Comment by laserkelvin** [comment]
@ys-teh unfortunately it seems like line numbers aren't available for the comments you made and I left unresolved...

**Comment by ys-teh** [comment]
> @ys-teh unfortunately it seems like line numbers aren't available for the comments you made and I left unresolved...

May I check which comment this refers to?

**Review by ys-teh — APPROVED** [review]
Approved, as most items have been resolved. Remember to address the remaining minor comments about duplicated lines. Thanks for taking the time to fix the documentation in the entire repo!
